input-profile.ts

```diff
@@ -0,0 +1,369 @@
+     * @hidden
+     */
+    private _setComponentsActive(active: boolean) {
+        const comps = this._defaultControllerComponents;
+        if (comps == undefined) return;
+        for (let i = 0; i < comps.length; ++i) {
+            comps[i].active = active;
+        }
+    }
+
+    /**
+     * Event handler triggered when XR input sources change.
+     * Detects new XR input sources and initiates the loading of input profiles.
+     * @param {XRInputSourceChangeEvent} event - The XR input source change event.
+     * @hidden
+     */
+    private _onInputSourcesChange(event: XRInputSourceChangeEvent) {
+        if (this._isModelLoaded() && !this.mapToDefaultController) {
+            this._setComponentsActive(false);
+        }
+
+        event.added.forEach((xrInputSource: XRInputSource) => {
+            if (xrInputSource.hand != null) return;
+            if (this._handedness != xrInputSource.handedness) return;
+
+            this.gamepad = xrInputSource.gamepad;
+            const profile =
+                this.customProfileFolder !== ''
+                    ? this.customProfileFolder
+                    : this.path + xrInputSource.profiles[0];
+            this.url = profile + '/profile.json';
+
+            this._profileJSON = InputProfile.Cache.has(this.url)
+                ? InputProfile.Cache.get(this.url)
+                : null;
+
+            if (this._profileJSON != null) return;
+            fetch(this.url)
+                .then((res) => res.json())
+                .then((out) => {
+                    this._profileJSON = out;
+                    InputProfile.Cache.set(this.url, this._profileJSON);
+
+                    if (!this._isModelLoaded()) this._loadAndMapGamepad(profile);
+                })
+                .catch((e) => {
+                    console.error('failed to load profile from' + this.url);
+                    console.error(e);
+                });
+        });
+    }
+
+    /**
+     * Checks if the 3D controller model is loaded.
+     * @return {boolean} True if the model is loaded; otherwise, false.
+     * @hidden
+     */
+    private _isModelLoaded() {
+        return this._controllerModel !== null;
+    }
+
+    /**
+     * Loads the 3D controller model and caches the mapping to the gamepad.
+     * @param {string} profile - The path to the input profile.
+     * @hidden
+     */
+    private async _loadAndMapGamepad(profile: string) {
+        const assetPath = profile + '/' + this._handedness + '.glb';
+        this._controllerModel = this.defaultController;
+        if (!this.mapToDefaultController) {
+            /** load 3d model in the runtime with profile url */
+            try {
+                this._controllerModel = (await this.engine.scene.append(
+                    assetPath
+                )) as Object3D;
+            } catch (e) {
+                console.error('failed to load 3d model');
+                console.error(e);
```

Comment:
```suggestion
                console.error('Failed to load 3d model', e);
```

And elsewhere (Captialize message start, and print the error in the same line).